Documentation/design-docs/standalone-gc-eventing.md

```diff
@@ -0,0 +1,212 @@
+void IGCToCLR::FireCustomEvent(
+    /* IN */ const char* eventName,
+    /* IN */ uint64_t eventId,
+    /* IN */ void* payload,
+    /* IN */ size_t payloadSize
+);
+```
+
+This callback fires a "custom" event that the EE is not previously aware of. The four arguments of this
+function are:
+
+1. `eventName` - The name of this custom event.
+2. `eventId` - A unique integer ID of this custom event that consumers can use to identify this event.
+3. `payload` - A raw binary payload of the event, passed verbatim to the underlying event implementation.
+4. `payloadSize` - The size of the binary payload.
+
+A runtime implementing this callback will implement it by having a "catch-all" GC event whose schema is
+an arbitrary sequence of bytes. Tools can parse the (deliberately unspecified) binary format provided by GC
+events that use this mechanism in order to recover the data within the payload.
+
+## Handling EE-specific arguments for GC events
+
+A number of events that the GC fires rely on state that is maintained by the EE. Several examples of this
+are:
+
+* Tracing controllers can request specific GC sequence numbers for the `GcStart` event. Before firing `GcStart`,
+the GC must first [obtain the requested sequence number](https://github.com/dotnet/coreclr/blob/46dc37a27abe79d5977a1942541b1665b61f7a17/src/vm/eventtrace.cpp#L894-L902).
+* Almost all events get the CLR instance ID (`GetClrInstanceId()`) and pass it as an argument to the
+event being fired.
+* Some events gather metadata about objects that are only known to the EE. Some examples of this would be
+[string representations of types](https://github.com/dotnet/coreclr/blob/46dc37a27abe79d5977a1942541b1665b61f7a17/src/gc/gcee.cpp#L468-L480) 
+("type names") and [app domains](https://github.com/dotnet/coreclr/blob/46dc37a27abe79d5977a1942541b1665b61f7a17/src/gc/handletable.cpp#L620).
+
+There are two concerns here that are at odds with one another. While we want to ensure that the API surface
+area of `IGCToCLR` is as minimal as possible, we also want to allow for custom events to have the ability
+to fire events whose payloads contain EE-known details. To this end I propose that we examine these bits
+of EE state leakage on a case-by-case basis. For EE state that is likely to be something that a custom event
+writer would use (for example - type names), callbacks on `IGCToCLR` are added to ensure that the GC has
+access to these details. For EE state that is not likely to be something used in custom events, we can
+"intercept" events as the flow from the GC to the EE but before they flow into the underlying event pipelines
+and add the EE-specific data at that point.
+
+To be more specific, the implementation of `IGCToCLR::FireKnownEvent` will look something like this:
+
+```c++
+void GCToEEInterface::FireKnownEvent(...)
+{
+    switch (eventId)
+    {
+        ...
+        case GcStartEventId:
+        {
+            LONGLONG l64ClientSequenceNumberToLog = GetClientSequenceNumber();
```

Comment:
Depending on how commonly this occurs, this pattern actually argues for what @jkotas was suggesting - a tracing interface with a single function per event.  Then, you can special case these easily and fill in the data that the GC doesn't have on the EE side right before the event is emitted.